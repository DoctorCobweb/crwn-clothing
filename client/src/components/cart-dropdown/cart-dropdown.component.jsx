import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import CartItem from '../cart-item/cart-item.component';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import {
  CartDropdownContainer,
  CartItemsContainer,
  EmptyMessageContainer,
  CartDropdownButton
} from './cart-dropdown.styles.jsx';

// see note below about default dispatch argument
export const CartDropdown = ({ cartItems, history, dispatch }) => (
  <CartDropdownContainer>
    <CartItemsContainer>
      {
        cartItems.length ? 
        cartItems.map(cartItem => (
          <CartItem key={cartItem.id} item={cartItem} />
        ))
        :
        <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
      }
    </CartItemsContainer>
    <CartDropdownButton
      onClick={() => {
        history.push('/checkout');
        dispatch(toggleCartHidden());
      }}
    >
      GO TO CHECKOUT
    </CartDropdownButton>
  </CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
});

// NOTE:
// this swapped order does NOT work. you don't get the history, match and path properties
// inside of mapStateToProps 
// export default connect(mapStateToProps, null)(withRouter(CartDropdown))

// DEFAULT dispatch ARGUMENT
// if you DONT add a mapDispatchToProps, redux actually sends in dispatch
// to props anyway. this allows for shorthand 'one-off' calls to
// dispatch if you need it, and not having to code up a separator
// mapDispatchToProps for some simple action.
export default withRouter(connect(mapStateToProps, null)(CartDropdown));