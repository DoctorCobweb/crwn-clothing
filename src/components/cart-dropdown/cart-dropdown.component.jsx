import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import CustomButton from '../custom-button/custom-button.component';
import CartItem from '../cart-item/cart-item.component';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import './cart-dropdown.styles.scss';

// see note below about default dispatch argument
const CartDropdown = ({ cartItems, history, dispatch }) => (
  <div className="cart-dropdown">
    <div className="cart-items" >
      {
        cartItems.length ? 
        cartItems.map(cartItem => (
          <CartItem key={cartItem.id} item={cartItem} />
        ))
        :
        <span className="empty-message">Your cart is empty</span>
      }
    </div>
    <CustomButton
      onClick={() => {
        history.push('/checkout');
        dispatch(toggleCartHidden());
      }}
    >
      GO TO CHECKOUT
    </CustomButton>
  </div>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
});


// this swapped order also seems to work ?!
// export default connect(mapStateToProps, null)(withRouter(CartDropdown))

// DEFAULT dispatch ARGUMENT
// if you DONT add a mapDispatchToProps, redux actually sends in dispatch
// to props anyway. this allows for shorthand 'one-off' calls to
// dispatch if you need it, and not having to code up a separator
// mapDispatchToProps for some simple action.
export default withRouter(connect(mapStateToProps, null)(CartDropdown));