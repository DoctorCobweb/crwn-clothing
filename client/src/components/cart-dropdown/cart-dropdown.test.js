import React from 'react';
import { shallow } from 'enzyme';
import { CartDropdown } from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

describe('CartDropdown component tests', ()=> {
  let mockCartItems;
  let mockHistory;
  let mockDispatch;
  let wrapper;

  beforeEach(() => {
    mockCartItems = [{id:1}, {id:2}, {id:3}]
    mockHistory = {
      push: jest.fn(),
    }
    mockDispatch = jest.fn();
    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch,
    }
    wrapper = shallow( <CartDropdown {...mockProps} />)
  })

  it('CartDropdown should match snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  })

  it('should call history.push when button is clicked', () => {
    wrapper.find('CartDropdownButton').simulate('click')
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockHistory.push).toHaveBeenCalledWith('/checkout');
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  })

  it('should render an equal number of CartItem components as the cartItems prop', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  })

  it('should render EmptyMessageContainer is cartItems is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch,
    }
    const newWrapper = shallow( <CartDropdown {...mockProps} />)

    // expect(newWrapper.find('EmptyMessageContainer').length).toEqual(1);
    expect(newWrapper.exists('EmptyMessageContainer')).toBe(true);
  })

})