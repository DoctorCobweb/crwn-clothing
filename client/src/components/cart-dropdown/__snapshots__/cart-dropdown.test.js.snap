// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component tests CartDropdown should match snapshot 1`] = `
"<CartDropdownContainer>
  <CartItemsContainer>
    <Memo(CartItem) item={{...}} />
    <Memo(CartItem) item={{...}} />
    <Memo(CartItem) item={{...}} />
  </CartItemsContainer>
  <CartDropdownButton onClick={[Function: onClick]}>
    GO TO CHECKOUT
  </CartDropdownButton>
</CartDropdownContainer>"
`;
